docs/tutorials/writing_a_new_data_struct.md

```diff
@@ -101,4 +105,202 @@ If everything is hooked together properly, JSON files for your new data struct s
+
+    fn try_from(other: &cldr_serde::numbers_json::Numbers) -> Result<Self, Self::Error> {
+        // This is the core transform operation. This step could take a lot of
+        // work, such as pre-parsing patterns, re-organizing the data, etc.
+    }
+}
+```
+
+The above example is an abridged snippet of code illustrating the most important boilerplate for implementing and ICU4X data transform.
+
+### `CldrJsonDataProvider`
+
+New CLDR JSON transformers need to be discoverable from `CldrJsonDataProvider`. To do this, edit *provider/cldr/src/transform/mod.rs* and fill in your data provider in every function, struct, and match. For example:
+
+```rust
+pub fn get_all_cldr_keys() -> Vec<ResourceKey> {
+    // ...
+    result.extend(&numbers::ALL_KEYS);
+    // ...
+}
+
+#[derive(Debug)]
+pub struct CldrJsonDataProvider<'a, 'data> {
+    // ...
+    numbers: LazyCldrProvider<NumbersProvider>,
+    // ...
+}
+
+impl<'a> CldrJsonDataProvider<'a, '_> {
+    pub fn new(cldr_paths: &'a dyn CldrPaths) -> Self {
+        CldrJsonDataProvider {
+            // ...
+            numbers: Default::default(),
+            // ...
+        }
+    }
+}
+
+impl<'a, 'data> DataProvider<'data, SerdeSeDataStructMarker> for CldrJsonDataProvider<'a, 'data> {
+    fn load_payload(
+        &self,
+        req: &DataRequest,
+    ) -> Result<DataResponse<'data, SerdeSeDataStructMarker>, DataError> {
+        // ...
+        if let Some(result) = self.numbers.try_load_serde(req, self.cldr_paths)? {
+            return Ok(result);
+        }
+        // ...
+    }
+}
+
+impl<'a> IterableDataProviderCore for CldrJsonDataProvider<'a, '_> {
+    fn supported_options_for_key(
+        &self,
+        resc_key: &ResourceKey,
+    ) -> Result<Box<dyn Iterator<Item = ResourceOptions> + '_>, DataError> {
+        // ...
+        if let Some(resp) = self
+            .numbers
+            .try_supported_options(resc_key, self.cldr_paths)?
+        {
+            return Ok(Box::new(resp.into_iter()));
+        }
+        // ...
+    }
+}
+```
+
+### Data Generation Tool
+
+```rust
+fn export_cldr<'data>(
+    matches: &ArgMatches,
+    exporter: &mut (impl DataExporter<'data, SerdeSeDataStructMarker> + ?Sized),
+    allowed_locales: Option<&[LanguageIdentifier]>,
+) -> anyhow::Result<()> {
+    // Create the provider after parsing options (not shown):
+    let raw_provider = CldrJsonDataProvider::new(cldr_paths.as_ref());
+
+    // Apply locale filters (not shown):
+    let provider: &dyn IterableDataProvider<SerdeSeDataStructMarker>;
+
+    // Dump data from the source provider to the data exporter:
+    for key in keys.iter() {
+        log::info!("Writing key: {}", key);
+        icu_provider::export::export_from_iterable(key, provider, exporter)?;
+    }
+
+    Ok(())
+}
+```
+
+The above example is a snippet from `icu4x-datagen` illustrating how `CldrJsonDataProvider` ties in with the data generation tool. Most contributors will not need to touch this part of the code.
```

Comment:
Question: When should a contributor add new code to Data Generation Tool? I'm trying to understand the purpose of `icu4x-datagen`. Is it responsible for transforming "Source data Struct" to "ICU4X Data Struct" or serializing "ICU4X Data Struct" to "ICU4X JSON/Blob"?

Reply:
Clarified in the doc.  It does steps 2-4.